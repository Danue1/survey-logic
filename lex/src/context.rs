pub struct Context<'lex> {
    source: &'lex str,
    index: usize,
}

impl<'lex> Context<'lex> {
    pub const fn new(source: &'lex str) -> Self {
        Self { source, index: 0 }
    }

    pub const fn index(&self) -> usize {
        self.index
    }

    pub fn peek(&self) -> Option<char> {
        self.source.get(self.index..).and_then(|s| s.chars().next())
    }

    pub fn bump(&mut self) -> Option<char> {
        let c = self.peek()?;
        self.advance();
        Some(c)
    }

    #[inline]
    pub fn advance(&mut self) {
        self.index += 1;
    }

    #[inline]
    pub fn slice(&self, start: usize) -> String {
        self.source[start..self.index].to_owned()
    }
}
