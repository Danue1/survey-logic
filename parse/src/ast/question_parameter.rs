use crate::context::{Context, State};
use syntax_kind::SyntaxKind;

impl Context {
    pub fn question_parameter(&mut self) -> State {
        if self.peek_str().to_lowercase().as_str() != "q" {
            return State::Continue;
        }

        self.start_node(SyntaxKind::QUESTION_PARAMETER_NODE);
        self.bump();
        self.numeric_literal();
        self.finish_node();
        State::Stop
    }
}

#[cfg(test)]
mod tests {
    macro_rules! assert_snapshot {
        ($parse:ident($source:expr)) => {
            let tokens = lex::lex($source);
            let mut context = super::Context::new(tokens);
            context.$parse();
            let node = context.finish();
            insta::assert_debug_snapshot!(node);
        };
    }

    #[test]
    fn question_parameter() {
        assert_snapshot!(question_parameter("Q123"));
    }
}
