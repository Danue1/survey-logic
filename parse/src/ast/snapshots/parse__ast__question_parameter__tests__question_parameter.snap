---
source: parse/src/ast/question_parameter.rs
expression: node
---
QUESTION_PARAMETER_NODE@0..4
  IDENTIFIER@0..1 "Q"
  NUMERIC_LITERAL_NODE@1..4
    NUMERIC@1..4 "123"

